components/scream/src/share/tests/CMakeLists.txt

```diff
@@ -1,22 +1,19 @@
 INCLUDE (ScreamUtils)
 
-SET (num_mpi_ranks 1)
-SET (config_defines "")
-
 # Test utilities (f90)
-CreateUnitTest(array_io "array_io_tests.cpp;unit_test.f90" scream_share num_mpi_ranks config_defines)
+CreateUnitTest(array_io "array_io_tests.cpp;unit_test.f90" scream_share)
 
 # Test utilities (c++)
-CreateUnitTest(util_cxx util_tests.cpp scream_share num_mpi_ranks config_defines)
+CreateUnitTest(util_cxx util_tests.cpp scream_share)
 
 # Test packs
-CreateUnitTest(packs "pack_tests.cpp;pack_kokkos_tests.cpp" scream_share num_mpi_ranks config_defines)
+CreateUnitTest(packs "pack_tests.cpp;pack_kokkos_tests.cpp" scream_share)
 
 # Test kokkos utils
-CreateUnitTest(kokkos_utils "kokkos_utils_tests.cpp" scream_share num_mpi_ranks config_defines)
+CreateUnitTest(kokkos_utils "kokkos_utils_tests.cpp" scream_share THREADS 1 16)
 
 # Test fields
-CreateUnitTest(field "field_tests.cpp" scream_share num_mpi_ranks config_defines)
+CreateUnitTest(field "field_tests.cpp" scream_share)
 
 # Test workspace manager
-CreateUnitTest(wsm "workspace_tests.cpp" scream_share num_mpi_ranks config_defines)
+CreateUnitTest(wsm "workspace_tests.cpp" scream_share THREADS 1 16)
```

Comment:
I'm thinking this will produce 16 separate unit tests. I think it would be better to produce just 1, like in scream-docs, that sweeps over multiple thread counts. Ideally, on KNL we could so a sweep of something like nthread = 1 to 272 stepping by 11, all in one unit test.

Reply:
> I'm thinking this will produce 16 separate unit tests.
Yes, it will.

I think if I can make the MAX_THREADS and the step more flexible, then this is a better approach than doing it like it's being done in scream-docs. Here's is my reasoning:

1. It's the clearest way to show the user which thread/rank combos are broken
2. It's nicer to watch 10 3s tests finish than getting no output for 30s
3. The current setup of scream_session driving catch2 tests is not conducive to the approach in the micro-app. I would have to make a different scream_session that accepted command-line arguments for thread configs.
4. Mpi-rank stepping definitely cannot be done within a single process, so this approach is necessary, so why not leverage it for threads too?

The one drawback is that, if the user wants to switch between thread counts/steps, they have to re-run cmake (once I expose SCREAM_TEST_MAX_THREAD etc) or write a custom bash loop.

Reply:
@bartgol , feel free to weigh-in here. This is a somewhat impactful decision-point for us.

Reply:
So fundamentally the test exe will be controlled by OMP_NUM_THREADS and mpirun flags? That's fine in the sense that explicit control is still easy. But see the final point about ctest output footprint.

Re: 30s testing. I think that's due to the equivalent of the scream-docs unittest_workspace test. Given unittest_team_utils, the workspace test can run with just the max threads. That will speed up testing quite a bit. (I put the 'brief' option in scream-docs so I could do this when using the test driver to drive dev outside of ctest.)

@bartgol do you find 1 ctest per thread count OK? I'm resistant b/c I feel it adds a lot of footprint to the ctest output. But I'm fine with being outvoted.

Reply:
I have no strong opinion about it. I also like short tests. I am not really concerned about ctest output (maybe I'm missing some issues related to it?).

In this case I would probably go for the easiest way to implement it. If Jim's solution is already working and without any potential issue related to it (that we can think of), then I would go with it. We can always roll back to a single test with multiple thread configurations later, no?

The concern about cmake's might be non negligible. So far I guess we are ok, since testbeds support it, but we should check what's on summit and cori, before blessing it as the ultimate solution.